examples/gno.land/r/gnoland/valopers/valopers.gno

```diff
@@ -0,0 +1,468 @@
+	if address == newAddress {
+		panic(ErrSameAddress)
+	}
+
+	v := GetByAddr(address)
+
+	// Check that the caller is the valoper
+	if v.Caller != std.PreviousRealm().Address() {
+		panic(ErrOriginCaller)
+	}
+
+	// Check that the address is valid
+	if err := validateBech32(newAddress.String()); err != nil {
+		panic(ErrInvalidAddress)
+	}
+
+	// Check that the new address wouldn't be
+	// overwriting an existing one
+	if isValoper(newAddress) {
+		panic(ErrAddressExists)
+	}
+
+	// Update the address
+	v.Address = newAddress
+
+	// Remove the old valoper info
+	valopers.Remove(address.String())
+
+	// Save the new valoper info
+	valopers.Set(newAddress.String(), v)
+}
+
+// UpdatePubKey updates an existing valoper's public key
+func UpdatePubKey(address std.Address, pubKey string) {
+	v := GetByAddr(address)
+
+	// Check that the caller is the valoper
+	if v.Caller != std.PreviousRealm().Address() {
+		panic(ErrOriginCaller)
+	}
+
+	// Check that the pubKey is valid
+	if err := validatePubKey(pubKey); err != nil {
+		panic(ErrInvalidPubKey)
+	}
+
+	// Update the description
+	v.PubKey = pubKey
+
+	// Save the valoper info
+	valopers.Set(address.String(), v)
+}
+
+// UpdateActive updates an existing valoper's active status
+func UpdateActive(address std.Address, active bool) {
+	v := GetByAddr(address)
+
+	// Check that the caller is the valoper
+	if v.Caller != std.PreviousRealm().Address() {
+		panic(ErrOriginCaller)
+	}
+
+	// Update active status
+	v.Active = active
+
+	// Save the valoper info
+	valopers.Set(address.String(), v)
+}
+
+// UpdateP2PAddresses updates an existing valoper's P2P addresses
+// p2pAddresses is a comma-separated list of P2P addresses
+func UpdateP2PAddresses(address std.Address, p2pAddresses string) {
+	v := GetByAddr(address)
+
+	// Check that the caller is the valoper
+	if v.Caller != std.PreviousRealm().Address() {
+		panic(ErrOriginCaller)
+	}
+
+	// Split the P2P addresses
+	addresses := splitP2PAddresses(p2pAddresses)
+
+	if len(addresses) == 0 {
+		panic(ErrP2PAddressesMissing)
+	}
+
+	// Check addresses
+	for _, addr := range addresses {
+		if err := validateP2PAddress(addr); err != nil {
+			panic(err)
+		}
+	}
+
+	// Update the P2P addresses
+	v.P2PAddresses = addresses
+
+	// Save the valoper info
+	valopers.Set(address.String(), v)
+}
+
+// GetByAddr fetches the valoper using the address, if present
+func GetByAddr(address std.Address) Valoper {
+	valoperRaw, exists := valopers.Get(address.String())
+	if !exists {
+		panic(ErrValoperMissing)
+	}
+
+	return valoperRaw.(Valoper)
+}
+
+// Render renders the current valoper set.
+// "/r/gnoland/valopers" lists all valopers, paginated.
+// "/r/gnoland/valopers:addr" shows the detail for the valoper with the addr.
+func Render(fullPath string) string {
+	path, _ := splitPathAndQuery(fullPath)
+	if path == "" {
+		return renderHome(fullPath)
+	} else {
+		if len(path) < 2 || path[:2] != "g1" {
+			return "invalid address " + path
+		}
```

Comment:
i suggest you to give a look at gno.land/p/moul/realmpath

Reply:
Thanks. This lets us remove the helper `splitPathAndQuery`. https://github.com/gnolang/gno/pull/3830/commits/15cb2b96f0a069402b1cb7c1e71e6ca6d62c8430

Reply:
Done: https://github.com/gnolang/gno/pull/3830/commits/15cb2b96f0a069402b1cb7c1e71e6ca6d62c8430